addons/event_crm/models/event_crm.py

```diff
@@ -0,0 +1,108 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import fields, models
+from odoo.tools.safe_eval import safe_eval
+
+
+class EventCrmRule(models.Model):
+    """
+    When registrations are created, if they match rules, we create leads depending of the
+    lead_creation_basis and we link those leads to the registrations. The lead creation type
+    determines the number of leads to create: one per order or one per attendee.
+    A rule can have 4 types of restrictions:
+        - The company of the event link to the registration
+        - The type of the event
+        - The event itself
+        - A domain set on event.registration
+
+    When setting the rule you can choose multiple event types and/or one particular event, so
+    it's totally possible to have a rule on multiple event types and specified one event which
+    does not belong to this categories.
+    """
+    _name = "event.crm.rule"
+    _description = "Event CRM Rules"
+    _order = "sequence, id"
+
+    # Rule configuration fields
+    name = fields.Char('Rule Name', required=True, translate=True)
+    sequence = fields.Integer(default=10)
+    active = fields.Boolean('Active', default=True)
+    lead_creation_basis = fields.Selection([
+        ('attendee', 'Per Attendee'), ('order', 'Per Order')],
+        string='Create', default='attendee', required=True,
+        help='One per Attendee : One lead is created for each attendee created (default):\n'
+            '\te.g. : I go to the "Salon de l\'Auto" page and register for me and 3 friends, 4 attendees are created and 4 leads are created. [Rather B2C]\n'
+            'One per Order : We create a single lead per "registration" (="ticket batch"):\n'
+            '\te.g. : I go to the Odoo Experience page and register for me and 3 colleagues, 4 attendees are created but a single lead is created.  [Rather B2B]')
+    event_type_ids = fields.Many2many('event.type',
+        string='Event Categories',
+        help='Only the categories selected will be impacted by this rule.')
+    event_id = fields.Many2one('event.event',
+        string='Event', domain="[('company_id', 'in', [company_id or current_company_id, False])]",
+        help='Registrations to this event will be impacted by this rule.')
+    company_id = fields.Many2one('res.company', string='Company',
+        help="Only events linked to this company will be impacted by this rule. "
+            "If it is not set, there will be no restriction applied to the company.")
+    registration_filter_domain = fields.Text(string="Registrations Domain", default='[]', required=True,
+        help="Additional domain which filters new registrations for the application of the rule.")
```

Comment:
Additional is not really the correct word if the domain only can generate leads.
If I understand well, lead can be generated based on registrations that matches:
- event_type_ids OR
- event_id OR
- registration_filter_domain

About the company, is it also a 'stand alone' criteria to generate leads ?
What if I have only company_id set ? (and not the rest) all events from that company will generate leads ?

This is quite unclear :thinking: